### DISCUSSION TITLE
It it able to excute command when do post request ?

### DISCUSSION THREAD

**Discussion by Zowie0122** [vercel/vercel#4606] [Help] [answered]
Hi, we are currently students in a coding Bootcamp. 

We have a team project, using next.js. The project is a web app for testing React codes. Github  ☞. https://github.com/Team-Draconis/Dragon_Vercel    For testing functionality, we use the ' exec ' package to run the command (‘yarn test’ )and 'fs.writeFileSync' to store user’s codes on the server. 

When we deployed this portion on to Vercel, we received a 502 bad response with an error message: NO_STATUS_CODE_FROM_FUNCTION . Within our current knowledge level, we are not quite sure what caused this problem. We assume that it might be involved with Vercel security protection (running command on Vercel server) or our subscription is not enough. 

Would you please suggest what we should do in terms of deployment? Many thanks! :)

**Comment by styfle** [comment, marked as answer]
Hi @Zowie0122 

Thanks for trying to Next.js and Vercel!

Vercel is designed for Static and Jamstack deployments using Serverless Functions.

So there are a couple architectural decisions with your current app that will lead to problems with Vercel.

1. The [`express` dependency](https://github.com/Team-Draconis/Dragon_Vercel/blob/master/package.json#L31) is not designed for serverless environments. Instead use Next.js `/pages/api` directory (looks like you already have).
2. The [`exec` call](https://github.com/Team-Draconis/Dragon_Vercel/blob/d0fd1d16fc19dbaac6f12e3fd016469a0c442a0e/pages/api/testRunner.js#L19) is assuming yarn is available but it is not installed as a [dependency](https://github.com/Team-Draconis/Dragon_Vercel/blob/master/package.json#L15). Furthermore, you would need to use it programatically instead of executing the CLI. Something like `import yarn from 'yarn'`. A better solution is to read the `test` script an execute that instead so then you don't need yarn at all.
3. Vercel only has a single writeable directory on the filesystem `/tmp`. You must make any [fs.writeFileSync](https://github.com/Team-Draconis/Dragon_Vercel/blob/master/pages/api/testRunner.js#L14) calls to that which can be found via `os.tmpdir()`.
